### PR TITLE
feat(k8s): bump to v1beta4

### PR DESCRIPTION
**Pull request by Sh4d1**
Signed-off-by: Patrik Cyvoct <pcyvoct@scaleway.com>

### CODE REVIEW

**Comment by Sh4d1** [comment]
Wont compile without #377

File: scaleway/resource_k8s_cluster_beta.go [hunk 1]
```diff
@@ -72,6 +84,42 @@ func resourceScalewayK8SClusterBeta() *schema.Resource {
...
+					Schema: map[string]*schema.Schema{
+						"enable": {
+							Type:        schema.TypeBool,
+							Required:    true,
+							Description: "Enables the Kubernetes patch version auto upgrade",
+						},
+						"maintenance_window_start_hour": {
+							Type:        schema.TypeInt,
+							Required:    true,
+							Description: "Start hour of the 2 hour maintenance window",
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: nit: 2-hour

File: scaleway/resource_k8s_cluster_beta.go [hunk 2]
```diff
@@ -281,6 +339,25 @@ func resourceScalewayK8SClusterBetaCreate(d *schema.ResourceData, m interface{})
 
 	req.AutoscalerConfig = autoscalerReq
 
+	autoUpgradeEnable, okAutoUpgradeEnable := d.GetOk("auto_upgrade.0.enable")
+	autoUpgradeStartHour, okAutoUpgradeStartHour := d.GetOk("auto_upgrade.0.maintenance_window_start_hour")
+	autoUpgradeDay, okAutoUpgradeDay := d.GetOk("auto_upgrade.0.maintenance_window_day")
+
+	// check if all or 0 values are set
+	if (okAutoUpgradeDay || okAutoUpgradeEnable || okAutoUpgradeStartHour == true) && (okAutoUpgradeDay && okAutoUpgradeEnable && okAutoUpgradeStartHour == false) {
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: nit: remove ` == true`

Comment (kindermoumoute) [reply scaleway/resource_k8s_cluster_beta.go]: something is wrong with this condition `(a or b or c) and (a and b and !c)` is the same as `a and b and !c`

Comment (kindermoumoute) [reply scaleway/resource_k8s_cluster_beta.go]: reminder: https://www.tutorialspoint.com/go/go_operators_precedence.htm

File: scaleway/resource_k8s_cluster_beta.go [hunk 3]
```diff
@@ -416,9 +493,11 @@ func resourceScalewayK8SClusterBetaRead(d *schema.ResourceData, m interface{}) e
 	d.Set("apiserver_url", response.ClusterURL)
 	d.Set("wildcard_dns", response.DNSWildcard)
 	d.Set("status", response.Status.String())
+	d.Set("upgrade_available", response.UpgradeAvailable)
 
 	// autoscaler_config
 	d.Set("autoscaler_config", []map[string]interface{}{clusterAutoscalerConfigFlatten(response)})
+	d.Set("auto_upgrade", []map[string]interface{}{clusterAutoUpgradeFlatten(response)})
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: `clusterAutoUpgradeFlatten` should return a `[]map[string]interface{}`

File: scaleway/resource_k8s_cluster_beta_test.go [hunk 4]
```diff
@@ -6,7 +6,7 @@ import (
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta_test.go]: Add a test in 2 steps to check maintenance update works correctly

File: scaleway/resource_k8s_cluster_beta.go (hunk 3 above)
Comment (Sh4d1) [reply scaleway/resource_k8s_cluster_beta.go]: In taht case, so should clusterAutoscalerConfigFlatten?

Comment (kindermoumoute) [reply scaleway/resource_k8s_cluster_beta.go]: yes

File: scaleway/resource_k8s_cluster_beta.go [hunk 5]
```diff
@@ -344,7 +344,7 @@ func resourceScalewayK8SClusterBetaCreate(d *schema.ResourceData, m interface{})
 	autoUpgradeDay, okAutoUpgradeDay := d.GetOk("auto_upgrade.0.maintenance_window_day")
 
 	// check if all or 0 values are set
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: comment must explain more
> // check if either all or none of the auto upgrade attribute are set.
> // if one auto upgrade attribute is set, they all must be set.
> // if none is set, auto upgrade attributes will be computed.

File: scaleway/resource_k8s_cluster_beta.go [hunk 6]
```diff
@@ -344,7 +344,7 @@ func resourceScalewayK8SClusterBetaCreate(d *schema.ResourceData, m interface{})
...
-	if (okAutoUpgradeDay || okAutoUpgradeEnable || okAutoUpgradeStartHour == true) && (okAutoUpgradeDay && okAutoUpgradeEnable && okAutoUpgradeStartHour == false) {
+	if (okAutoUpgradeDay || okAutoUpgradeEnable || okAutoUpgradeStartHour) && !(okAutoUpgradeDay && okAutoUpgradeEnable && okAutoUpgradeStartHour) {
```
Comment (kindermoumoute) [inline scaleway/resource_k8s_cluster_beta.go]: Let's keep it simple:
```
// We either want all true (all set) or all false (none set), so they must be equal
okAutoUpgradeDay ==okAutoUpgradeEnable && okAutoUpgradeEnable ==okAutoUpgradeStartHour
```

File: scaleway/resource_k8s_cluster_beta.go [hunk 7]
```diff
@@ -72,6 +84,42 @@ func resourceScalewayK8SClusterBeta() *schema.Resource {
...
+					Schema: map[string]*schema.Schema{
+						"enable": {
+							Type:        schema.TypeBool,
+							Required:    true,
+							Description: "Enables the Kubernetes patch version auto upgrade",
+						},
+						"maintenance_window_start_hour": {
+							Type:        schema.TypeInt,
+							Required:    true,
+							Description: "Start hour of the 2-hour maintenance window",
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_cluster_beta.go]: Can you add a `ValidateFunc` here?

File: scaleway/resource_k8s_cluster_beta.go [hunk 8]
```diff
@@ -281,6 +339,28 @@ func resourceScalewayK8SClusterBetaCreate(d *schema.ResourceData, m interface{})
...
+
+	// check if either all or none of the auto upgrade attribute are set.
+	// if one auto upgrade attribute is set, they all must be set.
+	// if none is set, auto upgrade attributes will be computed.
+	// We either want all true (all set) or all false (none set), so they must be equal
+	if okAutoUpgradeDay == okAutoUpgradeEnable && okAutoUpgradeEnable == okAutoUpgradeStartHour {
+		return fmt.Errorf("all field or zero field of auto_upgrade must be set")
+	}
+
+	if okAutoUpgradeDay && okAutoUpgradeEnable && okAutoUpgradeStartHour == true {
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_cluster_beta.go]: Remove `== true`

File: scaleway/resource_k8s_cluster_beta.go [hunk 9]
```diff
@@ -281,6 +339,28 @@ func resourceScalewayK8SClusterBetaCreate(d *schema.ResourceData, m interface{})
...
 
+	autoUpgradeEnable, okAutoUpgradeEnable := d.GetOk("auto_upgrade.0.enable")
+	autoUpgradeStartHour, okAutoUpgradeStartHour := d.GetOk("auto_upgrade.0.maintenance_window_start_hour")
+	autoUpgradeDay, okAutoUpgradeDay := d.GetOk("auto_upgrade.0.maintenance_window_day")
+
+	// check if either all or none of the auto upgrade attribute are set.
+	// if one auto upgrade attribute is set, they all must be set.
+	// if none is set, auto upgrade attributes will be computed.
+	// We either want all true (all set) or all false (none set), so they must be equal
+	if okAutoUpgradeDay == okAutoUpgradeEnable && okAutoUpgradeEnable == okAutoUpgradeStartHour {
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_cluster_beta.go]: Do we need to check that `autoUpgradeEnable == true` here?

nit: `if okAutoUpgradeEnable == okAutoUpgradeDay  && okAutoUpgradeEnable == okAutoUpgradeStartHour {`

File: scaleway/resource_k8s_cluster_beta.go [hunk 10]
```diff
@@ -281,6 +339,28 @@ func resourceScalewayK8SClusterBetaCreate(d *schema.ResourceData, m interface{})
 
 	req.AutoscalerConfig = autoscalerReq
 
+	autoUpgradeEnable, okAutoUpgradeEnable := d.GetOk("auto_upgrade.0.enable")
+	autoUpgradeStartHour, okAutoUpgradeStartHour := d.GetOk("auto_upgrade.0.maintenance_window_start_hour")
+	autoUpgradeDay, okAutoUpgradeDay := d.GetOk("auto_upgrade.0.maintenance_window_day")
+
+	// check if either all or none of the auto upgrade attribute are set.
```
Comment (QuentinBrosse) [inline scaleway/resource_k8s_cluster_beta.go]: Can you rework this comment, please? Lines 1, 2 and 4 of your comment say more or less the same things.

Comment (Sh4d1) [reply scaleway/resource_k8s_cluster_beta.go]: @kindermoumoute https://github.com/terraform-providers/terraform-provider-scaleway/pull/376#discussion_r360577837 :sweat_smile:

File: scaleway/resource_k8s_cluster_beta.go (hunk 9 above)
Comment (Sh4d1) [reply scaleway/resource_k8s_cluster_beta.go]: Nop, no need to check, it's done just after

File: scaleway/resource_k8s_cluster_beta.go (hunk 10 above)
Comment (Sh4d1) [reply scaleway/resource_k8s_cluster_beta.go]: removed the 4th line, should be better

Comment (Sh4d1) [reply scaleway/resource_k8s_cluster_beta.go]: removed the 4th line, should be better

**Pull request merged**